tsl/src/compression/api.c

```diff
@@ -396,8 +396,13 @@ decompress_chunk_impl(Oid uncompressed_hypertable_relid, Oid uncompressed_chunk_
 	/* Recreate FK constraints, since they were dropped during compression. */
 	ts_chunk_create_fks(uncompressed_chunk);
 
-	/* Prevent readers from using the compressed chunk that is going to be deleted */
-	LockRelationOid(uncompressed_chunk->table_id, AccessExclusiveLock);
+	/*
+	 * Prevents readers from using the compressed chunk that is going to be
+	 * deleted. Calling performMultipleDeletions in chunk_index_tuple_delete
+	 * also requests an AccessExclusiveLock. However, this call makes the
+	 * lock on the chunk explicit.
```

Comment:
This change not only makes the lock explicit, but also moves it earlier, before `ts_compression_chunk_size_delete`. Do we have a test case that breaks if we remove the explicit lock altogether?

Reply:
At the first glance, nothing should break because w/o the lock the concurrent SELECTs just stop seeing the compressed chunk during planning, earlier than it is actually dropped.

Reply:
I am not able to create a failing test for this case. However, I am in favor of making the lock explicit because also PostgreSQL requests an `AccessExclusiveLock` before `performMultipleDeletions` is [invoked](https://github.com/postgres/postgres/blob/b0a55e43299c4ea2a9a8c757f9c26352407d0ccc/src/backend/commands/tablecmds.c#L1335). So, I want to be compliant with PostgreSQL logic.

Reply:
But this is not only about explicitness, this is changing the relative order of events -- e.g. whether we modify the chunks table holding the exclusive lock on compressed chunk or not. If you want just to make it analogous to posgres, the place to take the deletion lock would be just before the `ts_chunk_drop` call, or maybe inside the `ts_chunk_drop_internal` call, or just relying on the `performDelete` doing the same thing.

If we explicitly want to lock it before modifying the catalog, I think we need some justification for this. I mean, if we have another problem in this place, someone will look at this lock again and think: why is it here and not there?

Given that the catalog modification and dropping the chunk happen inside a transaction, they are both going to become visible atomically, anyway, no matter where we put this lock.

Reply:
@akzum Moving the lock just before `ts_chunk_drop` is a good point to ensure the order of events is not changed.

Are you generally in favor of performing this lock or do you argue that it is not needed because we can rely on `performDelete`?

Reply:
> the order of events is not changed

Of which events exactly?

I think the locking performed by `performDelete` is enough for correctness. The reason I'm asking is that I don't understand the rationale for 1) taking this lock explicitly (well, maybe documentation and following the general postgres line of thought)
2) most importantly, the rationale for doing it at this particular line and not in `ts_chunk_drop_internal` like point (1) would suggest. If we are trying to prevent some unwanted sequence of events, what is this sequence exactly? I think it's important to have such a description in the future when we are going to modify this code, so that we are able to do it correctly.

Reply:
> Of which events exactly?

I mean, we probably should have a comment describing what the order is, and why it is important. At the first glance, there's no meaningful ordering of cleaning up the chunk record, the chunk size record and dropping the compressed relation -- to other transactions they happen atomically at the end of decompressing transaction.